src/invariant/test/OrderBookIsNotCrossedTests.cpp

```diff
@@ -0,0 +1,361 @@
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    OfferEntryCmp cmp;
+    OfferEntry l, r;
+
+    l.price = Price{1, 1};
+    l.flags = PASSIVE_FLAG;
+    r.price = Price{2, 1};
+    r.flags = 0;
+    REQUIRE(cmp(l, r));
+    REQUIRE(!cmp(r, l));
+
+    l.price = Price{1, 1};
+    l.flags = 0;
+    l.offerID = 2;
+    r.price = Price{1, 1};
+    r.flags = PASSIVE_FLAG;
+    r.offerID = 1;
+    REQUIRE(cmp(l, r));
+    REQUIRE(!cmp(r, l));
+}
+
+TEST_CASE("OrderBookIsNotCrossed in-memory order book is consistent with "
+          "application behaviour",
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    Asset cur1;
+    cur1.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur1.alphaNum4().assetCode, "CUR1");
+
+    Asset cur2;
+    cur2.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur2.alphaNum4().assetCode, "CUR2");
+
+    VirtualClock clock;
+    auto cfg = getTestConfig(0);
+    // When testing the order book not crossed invariant, enable it and no other
+    // invariants (these tests do things which violate other invariants).
+    cfg.INVARIANT_CHECKS = {};
+    auto app = createTestApplication(clock, cfg);
+    OrderBookIsNotCrossed::registerAndEnableInvariant(*app);
+    app->start();
+
+    LedgerTxn ltxOuter{app->getLedgerTxnRoot()};
+
+    auto const invariant = std::make_shared<OrderBookIsNotCrossed>();
+    auto offer = InvariantTestUtils::generateOffer(cur1, cur2, 3, Price{3, 2});
+
+    // create
+    {
+        LedgerTxn ltx{ltxOuter};
+        ltx.create(offer);
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+        auto const& orders = invariant->getOrderBook().at({cur1, cur2});
+
+        REQUIRE(orders.size() == 1);
+
+        auto const& offerToCheck = *orders.cbegin();
+
+        REQUIRE(offerToCheck.amount == 3);
+        REQUIRE(offerToCheck.price.n == 3);
+        REQUIRE(offerToCheck.price.d == 2);
+
+        ltx.commit();
+    }
+
+    // modify
+    {
+        LedgerTxn ltx{ltxOuter};
+
+        offer.data.offer().amount = 2;
+        offer.data.offer().price = Price{5, 3};
+        auto entry = ltx.load(LedgerEntryKey(offer));
+        entry.current() = offer;
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+        auto const& orders = invariant->getOrderBook().at({cur1, cur2});
+
+        REQUIRE(orders.size() == 1);
+
+        auto const& offerToCheck = *orders.cbegin();
+
+        REQUIRE(offerToCheck.amount == 2);
+        REQUIRE(offerToCheck.price.n == 5);
+
+        ltx.commit();
+    }
+
+    // delete
+    {
+        LedgerTxn ltx{ltxOuter};
+
+        auto entry = ltx.load(LedgerEntryKey(offer));
+        entry.erase();
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+
+        REQUIRE(invariant->getOrderBook().at({cur1, cur2}).size() == 0);
+    }
+}
+
+TEST_CASE("OrderBookIsNotCrossed properly throws if order book is crossed",
```

Comment:
One interesting group of cases that you aren't testing: changing the assets on an offer (this is permitted, you can even swap them). You might want to add some tests for this in both this case and the previous one.

Reply:
That _is_ interesting.  I added several tests.  It also made me realize that you could cross the book by making a passive offer non-passive, so I added a test for that too.